specs/dynamicdependencies/DynamicDependencies.md

```diff
@@ -0,0 +1,1363 @@
+        DetourTransactionCommit()
+    }
+    else if (dwReason == DLL_PROCESS_DETACH)
+    {
+        DetourTransactionBegin()
+        DetourUpdateThread(GetCurrentThread())
+        DetourDetach(&TrueGetCurrentPackageInfo, DynamicGetCurrentPackageInfo)
+        ...
+        DetourTransactionCommit()
+    }
+    return TRUE;
+}
+
+LONG DynamicGetCurrentPackageInfo(...PACKAGE_INFO* packageInfo...UINT32* count)
+{
+    *count = 0
+
+    std::unique_lock<std>>mutex> lock(g_lock)
+
+    for (node : g_packageGraph)
+    {
+        auto packageInfoReference = node.get()
+        if (node)
+        {
+            packageInfo += GetPackageInfo(packageInfoRefernce...)
+        }
+        else
+        {
+            packageInfo += TrueGetCurrentPackageInfo(...)
+        }
+        ++count
+    }
+    return ERROR_SUCCESS;
+}
+...
+
+HRESULT MddAddPackageDependency(...)
+{
+    ...
+
+    std::unique_lock<std>>mutex> lock(g_lock)
+}
+```
+
+### 3.1.2. Add to the Package Graph
+
+`MddAddPackageDependency` has 2 responsibilities:
+
+1. Resolve a package dependency to a specific package
+2. Update the package graph
+
+Updating the package graph also requires the DLL Search Order be updated. See below for more details.
+
+```c++ (C++ish pseudocode)
+HRESULT MddAddPackageDependency(...)
+{
+    std::unique_lock<std>>mutex> lock(g_lock)
+
+    auto packageFullName = ResolvePackageDependency(packageFamilyName, minVersion, architectureFilter, ...)
+    if (!packageFullName)
+    {
+        return ERROR
+    }
+    return AddToPackageGraph(packageFullName, ...)
+}
+```
+
+#### 3.1.2.1. Resolve Package
+
+Resolve the package dependency to a (specific) package, per the following conditions:
+
+1. Packages in Package Family
+2. PackageType = Framework
+3. Packages registered to the user
+4. Highest version
+5. Version >= packageDependency.MinVersion
+6. Architecture meets ArchitectureFilter criteria (see below)
+7. PackageStatus = usable (not Bad, not NotAvailable, etc)
+
+If 2+ meet the criteria due, then the best-fit is...
+
+1. Highest version
+2. Architecture exactly matches caller's architecture
+3. Architecture=Neutral
+
+Thus v5 is better than v4, but choosing amongst v5-x86 vs v5-x64 vs v5-neutral
+depends on the caller's architecture. For example, assuming multiple package
+candiates only varying by version and/or architecture:
+
+Given v1 and v2
+
+- Best-Fit = v2
+
+Given v1-x86 and v1-x64
+
+- Best-Fit for x64 callers = v1-x64
+- Best-Fit for x86 callers = v1-x86
+- Best-Fit for arm callers = --no match--
+
+Given v1-x86 and v1-neutral
+
+- Best-Fit for x64 callers = v1-neutral
+- Best-Fit for x86 callers = v1-x86
+- Best-Fit for arm callers = v1-neutral
+
+**NOTE:** `MddAddPackageDependency` can only resolve to packages registered for the user. Thus as
+packages cannot be registered to LocalSystem a process running as LocalSystem cannot use Dynamic
+Dependencies. Impersonation is irrelevant here. Only the process' user identity is relevant to
```

Comment:
I know we've had issues with elevated apps (e.g. Terminal) using framework packages today. Does this spec address that or is there some nuance to the problems that Terminal has hit?

Reply:
I'm not sure the elevation issues you're referring to. I'll look into it. Thanks Jevan!

Reply:
The elevation issues are a different matter having to do with user1 using packages registered to user2 (and not user1). Addressing that may require changes inside Windows, but regardless it's not something specific to Dynamic Dependencies. If this interests you please open a new issue for further discussion.